### DISCUSSION TITLE
trimesh and shapely objects are hasheable... so... (it allows some playful hacks)

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#419] [Show and tell]
You can use the `functools.cache' decorator on functions that get them as arguments, and it might allow you to convert them only when they change, otherwise returning a cached py5.Py5Shape object.

```python
import functools 

@functools.cache    # or @functools.lru_cache(maxsize=30) if you will recalculate the mesh a lot
def convert_obj(obj):
    return py5.convert_shape(obj)
```

**Comment by hx2A** [comment]
They are hashable but only Shapely 2.0 objects are [immutable](https://shapely.readthedocs.io/en/stable/migration.html#geometry-objects-will-become-immutable). Shapely <2.0 objects and Trimesh objects are not. Relying on `hash()` for mutable objects is not a good idea.

However, I see why you are asking this. Currently the user now has a chore to convert an object once with `convert_shape()` and then to save the output. Is there a way to relieve the user of this burden that is also safe and reliable? I don't know, but let's think about this and see what we can come up with.

**Reply by villares** [reply]
My reasoning is:

If I'm not planning to mutate the "final"  shapely object or the trimesh mesh that I want to render with py5, I found it handy to be able to repeatedly call my cached conversion function in draw, so I get a cached Py5Shape, and if I redo the object/mesh calculation due to an interactive event, I get a new object that most probably gets a new hash and will be converted again. 

The alternative I was using before was to have a flag to trigger a new conversion when I recalculated objects... 

Does that make sense?

**Reply by hx2A** [reply]
I see what you are saying. Not all objects are hashable though - trimesh and shapely objects are, but numpy arrays are not. The `convert_image()` method supports unhashable objects, and although `convert_shape()` currently does not, it probably will in the future.

I don't like the idea of silent caching for this because it could lead to memory problems if someone were to feed in a lot of new objects for every frame. How about instead creating new cache methods, such as `cached_convert_image()` and `cached_convert_shape()`, that have this extra functionality?

This approach is dependent on the other libraries' hash functions being comprehensive so that there are no ways to alter the object without also changing the hash. Given the complexity of the trimesh library I'm skeptical that that is the case. Some quick experiments seem to indicate that changing the mesh changes the hash, but it only takes one counterexample to prove this wrong. Also, as py5's conversion functionality grows, it's likely a new library will be added that gets it wrong. On the other hand, if the risk is documented and people are willing to manage the risk, it should be OK.

Maybe something like this could work:

```python
def cached_convert_shape(obj: Any, *, force_conversion=False):
    ...
```

Users pass `force_conversion=True` to force a re-conversion when they change something and don't want to rely on the hash function.

Also, that method would need to issue a one-time warning when passed an object that is not hashable.

**Reply by villares** [reply]
Oh, I guess I didn't explain myself well... you are thinking like the wise maintainer that you are, generalizing this and looking at the consequences of it, but this is not a proposal for new functionality in py5! I was just sharing a brittle hack I was applying just to shapely and trimesh on my exploratory sketches.  :D

**Reply by hx2A** [reply]
Ah, well, I've faced the same problem with needing to avoid repeatedly converting the same object, and I've been addressing it by converting in setup and storing the result in a global variable. This works, but it is a bit tedious when you have half a dozen objects to convert. A working caching mechanism would add a lot of value.

I want to think through some of the details of the design before implementing it so it won't go in the next release, but it will be in the release after the next one. I'll open an issue to track this.

**Reply by hx2A** [reply]
See #420

**Reply by villares** [reply]
> See #420 

If you worry about caching eating too much memory,  maybe you could use @lru_cache with a maxlen= ...

**Reply by hx2A** [reply]
If I set a fixed size there'd need to be another method for users to set the size. I was more concerned about automatically caching Java objects that would never be used again that overwhelm the JVM. I think the guidance should be to use `convert_shape()` if they are going to use the converted object once and use `cached_convert_shape()` if they used it multiple times, with the latter being the equivalent memory impact of calling `convert_shape()` in setup, storing in a global variable, etc.

**Reply by hx2A** [reply]
> This approach is dependent on the other libraries' hash functions being comprehensive so that there are no ways to alter the object without also changing the hash. Given the complexity of the trimesh library I'm skeptical that that is the case.

It seems the trimesh folks have thought carefully about hashing objects:

https://trimesh.org/trimesh.caching.html#trimesh.caching.TrackedArray
